File: serde-generate/runtime/typescript/serde/binarySerializer.ts
import {BigNumber} from '@ethersproject/bignumber';
import {Int64LE, Uint64LE} from 'int64-buffer';
import {Serializer} from './serializer';
import bytes from '@ethersproject/bytes';

export abstract class BinarySerializer implements Serializer {
    private output: Uint8Array = Buffer.alloc(0);

    abstract serializeLen(value: number): void ;

    abstract serializeVariantIndex(value: number): void;

    abstract sortMapEntries(offsets: number[]): void;

    public serializeStr(value: string): void {
        const stringUTF8Bytes = Buffer.from(value, 'utf8');
        this.serializeBytes(stringUTF8Bytes);
    }

    public serializeBytes(value: Uint8Array): void {
        this.serializeLen(value.length);
        this.concat(value);
    }

    public serializeBool(value: boolean): void {
        this.concat(Buffer.from([value ? 1 : 0]));
    }

    public serializeUnit(value: any): void {
    }

    public serializeU8(value: number): void {
        this.concat(new Uint8Array([value]));
    }

    public serializeU16(value: number): void {
        const u16 = new Uint16Array([value]);
        this.concat(Buffer.from(u16.buffer));
    }

    public serializeU32(value: number): void {
        const u32 = new Uint32Array([value]);
        this.concat(Buffer.from(u32.buffer));
    }

    public serializeU64(value: Uint64LE): void {
        const buffer = value.toArrayBuffer();
        this.concat(new Uint8Array(buffer));
    }

    public serializeU128(value: BigNumber): void {
        this.concat(bytes.arrayify(value));
    }

    public serializeI8(value: number): void {
        this.serializeU8(value);
    }

    public serializeI16(value: number): void {
        this.serializeU16(value);
    }

    public serializeI32(value: number): void {
        this.serializeU32(value);
    }

    public serializeI64(value: Int64LE): void {
        const buffer = value.toArrayBuffer();
        this.concat(new Uint8Array(buffer));
    }

    public serializeI128(value: BigNumber): void {
        this.concat(bytes.arrayify(value));
    }

    public serializeOptionTag(value: boolean): void {
        if (value) {
            this.concat(Buffer.from([1])); // True
        } else {
            this.concat(Buffer.from([0])); // False
        }
    }

    public getBufferOffset(): number {
        return this.output.length;
    }

    public getBytes(): Uint8Array {
        return this.output;
    }

    public concat(value: Uint8Array): void {
        this.output = BinarySerializer.concat(this.output, value);
    }

    public static concat(a: Uint8Array, b: Uint8Array): Uint8Array {
        return Buffer.concat([a, b], a.length + b.length);
    }

    public serializeChar(value: string): void {
        throw new Error('Method serializeChar not implemented.');
    }

    public serializeF32(value: number): void {
        throw new Error('Method serializeF32 not implemented.');
    }

    public serializeF64(value: number): void {
        throw new Error('Method serializeF64 not implemented.');
    }
}

File: serde-generate/runtime/swift/Serde/BinarySerializer.swift
//  Copyright © Diem Association. All rights reserved.

enum BinarySerializerError: Error {
  case serializationException(issue: String)
}

import Foundation

public class BinarySerializer : Serializer {
  let buffer: InputStream
  public var output = OutputStream.toMemory()
  private var containerDepthBudget:Int64
  
  init(maxContainerDepth:Int64) {
    buffer = InputStream()
    output.open()
    containerDepthBudget = maxContainerDepth;
  }
  
  deinit {
    output.close()
  }
  
  func increase_container_depth() throws {
    if containerDepthBudget == 0 {
      throw BinarySerializerError.serializationException(issue: "Exceeded maximum container depth")
    }
    containerDepthBudget -= 1
  }
  
  func decrease_container_depth() {
    containerDepthBudget += 1
  }
  
  func serialize_char(value: Character) throws {
    throw BinarySerializerError.serializationException(issue: "Not implemented: char serialization")
  }
  
  func serialize_f32(value: Float) {
    output.write(data: Data(fromArray:[value]))
  }
  
  func serialize_f64(value: Double) {
    output.write(data: Data(fromArray:[value]))
  }
  
  func get_bytes() -> [UInt8] {
    return output.getBuffer()
    //return [UInt8](Data(reading: buffer))
  }
  
  func serialize_str(value: String) throws {
    let buf: [UInt8] = Array(value.utf8)
    serialize_bytes(value: buf)
  }
  
  func serialize_bytes(value: [UInt8]) {
    serialize_len(value: Int64(value.count))
    for b in value {
      output.write(data: Data([b]))
    }
  }
  
  func serialize_bool(value: Bool) {
    var valueInInt = value == true ? 1 : 0
    let data = Data(bytes: &valueInInt, count: MemoryLayout.size(ofValue: valueInInt)) //Int to Data
    output.write(data: data)
  }
  
  func serialize_unit(value: Unit) {
    output.write(data: Data(fromArray:[value]))
  }
  
  func serialize_u8(value: UInt8) {
    output.write(data: Data(fromArray:[value]))
  }
  
  private func custom_write<T>(value: T) -> Void {
    var value = value
    let size = MemoryLayout.size(ofValue: value)
    withUnsafeBytes(of: &value) {
      ptr in
      output.write(ptr.baseAddress!.assumingMemoryBound(to: UInt8.self), maxLength: size)
    }
  }
  
  func serialize_u16(value: UInt16) {
    output.write(data: Data(fromArray: [value]))
  }
  
  func serialize_u32(value: UInt32) {
    output.write(data: Data(fromArray: [value]))
  }
  
  func serialize_u64(value: UInt64) {
    output.write(data: Data(fromArray: [value]))
  }
  
  func serialize_u128(value: BigInt8) throws {
    if value >> 128 != 0 {
      throw BinarySerializerError.serializationException(issue: "Invalid value for an unsigned int128")
    }
    
    assert(value._data.count <= 16 || value._data[16] == 0)
    
    for i in 0..<16 {
      if i < value._data.count {
        output.write(data: Data(fromArray: [value._data[i]]))
      } else {
        output.write(data: Data(fromArray: [UInt8(0)]))
      }
    }
  }
  
  func serialize_i8(value: Int8) {
    output.write(data: Data(fromArray: [value]))
    return print(value)
  }
  
  func serialize_i16(value: Int16) {
    output.write(data: Data(fromArray: [value]))
    return print(value)
  }
  
  func serialize_i32(value: Int32) {
    output.write(data: Data(fromArray: [value]))
    return print(value)
  }
  
  func serialize_i64(value: Int64) {
    output.write(data: Data(fromArray: [value]))
    return print(value)
  }
  
  func serialize_i128(value: BigInt8) throws {
    if value >= 0 {
      if value >> 127 != 0 {
        throw BinarySerializerError.serializationException(issue: "Invalid value for a signed int128")
      }
      try serialize_u128(value: value)
    }
    else {
      if -(value + 1) >> 127 != 0 {
        throw BinarySerializerError.serializationException(issue: "Invalid value for a signed int128")
      }
      try serialize_u128(value: value + (BigInt8(1) << 128))
    }
  }
  
  func serialize_option_tag(value: Bool) {
    output.write(data: Data(fromArray: [value]))
  }
  
  func get_buffer_offset() -> Int {
    return output.getOffset()
  }
  
  func serialize_len(value: Int64) {
    output.write(data: Data(fromArray: [value]))
  }
  
  func serialize_variant_index(value: Int) {
    output.write(data: Data(fromArray: [value]))
    print(value)
  }
  
  func sort_map_entries(offsets: [Int]) {
    // Not required by the format.
  }
  
  private func write<T>(value: T) {
    var value = value
    let size = MemoryLayout.size(ofValue: value)
    withUnsafeBytes(of: &value) {
      ptr in
      output.write(ptr.baseAddress!.assumingMemoryBound(to: UInt8.self), maxLength: size)
    }
  }
}
